### ISSUE TITLE
Async Multiplayer

### ISSUE THREAD

**Issue by boozook** [vcmi/vcmi#1220]
I'm strongly sure that async multiplayer mode more useful in most cases especially mobile devices and so should be implemented.

> In next words "save-game" means some transferable game data that contains changes of state - human player's turn.

_Async Multiplayer is when you can do your turn offline, then send the turn as save-game or state-diff somehow to other players. You should no to be connected to any server to play, just to send entire turn after all._

Additionally it can be base for native implementation on top of existing platform-dependent systems such as GameCenter or same on Android.

⚠️ One potential problem I see is some inter-turn interactions, that requires two players such as battle, exchange, etc.
If it possible to be solved, the implementation plan could looks like this:

- [x] ability to save and restore any game
- [ ] export, share save-game (so player can send it to another via email or messenger)
    - [ ] send to server (simple backend with mapping room->player-save-game)
    - [ ] sync with cloud (Dropbox, iCloud, etc)
- [ ] UI (mode, options)

**Comment by boozook** [comment]
Look what I've found!

[GKTurnBasedMatch.sendExchange] is for inter-turn interaction. Is I understand it right, it's exactly what we need for battles and item exchanges between heroes.

[GKTurnBasedMatch.sendExchange]: https://developer.apple.com/documentation/gamekit/gkturnbasedmatch/1520451-sendexchange

**Comment by kambala-decapitator** [comment]
@boozook not sure how that's gonna help because:
- it's for Apple platforms only
- and accessing Game Center requires VCMI to be present on AppStore

**Comment by boozook** [comment]
> - and accessing Game Center requires VCMI to be present on AppStore

Game Center doesn't require presenting in AppStore, but simple config page in Connect to enable GameCenter. So in this way we can continuously deploy build to TestFlight for "external testing".

**Comment by kambala-decapitator** [comment]
> > * and accessing Game Center requires VCMI to be present on AppStore
> 
> Game Center doesn't require presenting in AppStore, but simple config page in Connect to enable GameCenter. So in this way we can continuously deploy build to TestFlight for "external testing".

it means having paid developer account anyway. And uploading to TF would violate GPL in the same way as distributing on the AppStore, which some of core devs strongly care about.

**Comment by rilian-la-te** [comment]
@kambala-decapitator  and it will not work on Linux and Android anyway)

**Comment by boozook** [comment]
> it means having paid developer account anyway.

I have one.

> And uploading to TF would violate GPL

So, that is unsolvable problem, so this issue could be closed 🤷‍♂️🥺


> and iirc TF allows max 10k testers?

Yup.

**Comment by kambala-decapitator** [comment]
> I have one 

Ideally, vcmi organization should have its own :)

> So, that is unsolvable problem, so this issue could be closed

I hope that one day this stance would change

**Comment by rilian-la-te** [comment]
@kambala-decapitator I think than VCMI should remain GPL or LGPL, I do not like paid programs.

**Comment by rilian-la-te** [comment]
About async multiplayer - we can have PBEM, which is free for all.

**Comment by kambala-decapitator** [comment]
> @kambala-decapitator I think than VCMI should remain GPL or LGPL, I do not like paid programs. 

It's not about making a paid app but rather that the very fact of uploading to TF / appstore violates a few GPL license terms. Dual license would solve that though.

**Comment by dydzio0614** [comment]
Not exactly, it gonna be singleplayer tournament where sending saves to organizer will be used as semi-proof of playing the map properly + will give food for players to analyse the course of actions

**Comment by gamestales** [comment]
I would like to achieve this for friend playing on windows
I have done it with "Battle for Wesnoth" (it allows many runtime parameters)

It would like this, players share some network disk (google drive)
One of them setups the game - they communicate over external channel
They set up path to the save file - each of them on his own machine.
Game is started with savefile parameter and automaticly loads it - we already have testfile in VCMI. 
Upon turn end the game autosaves and exits.

It seem only few things need to be done.
1. Savefile parameter - similar to testfile
2. Singleturn parameter - so turn end would autosave and close the game 
3. Some script for running game with path to savefile

**I think I can do it if the idea is accepted.**

**Comment by IvanSavenko** [comment]
Copying from Slack:
- How to handle player-vs-player battles? We don't support saves during battles and even if we do, overhead from sending saves would make battles extremely slow.
- How and where we should make UI, even something as "simple" as selecting starting town, without network game? Or game chat?
- Do we need such feature? It is very niche with no clear use-case. Besides, pvp in HoMM is usually based on simturns, not such "play by email" mode. I really don't see any audience for such feature.

**Comment by IvanSavenko** [comment]
For such cases people can use alt-tab on desktop or send app into background on mobiles.

And considering current state of multiplayer code I am rather reluctant to add more code in there that we need to support.
If we can implement abstraction layer for connections and allow different "connection" types, e.g. same process / interprocess / network TCP / network UDP then something like this can be considered. 

But as of right now yeah, too low priority IMO to even maintain this code IMO.

**Comment by gamestales** [comment]
The full async multiplier would indeed require much work. 
I talk about very small changes, would need 3 start parameters to make running "correspondence" game smoothly.
Start game from save file, end game it after turn end.
All else I would do with external scripting.

**Comment by IvanSavenko** [comment]
And how to handle pvp battles for example?

Also - if goal is to support mobile devices then how are you going to run external scripting on Android or iOS where most of inter-process access is banned and there is no way to run scripts without rooting device?

Not to mention that it is 2023, not 2003. Pretty much everyone has 24/7 access to Internet on their phone.

**Comment by kraemerb** [comment]
I'm new here and just discovered this chat, but I just want to say that asynchronous multiplayer for mobile is by far the feature I would be grateful for more than any other. For those who indicated that this is a niche/low priority request, you may be right that it may not be used as much among the hardcore, most active members of this community. However, I think this feature, if executed well, could be HUGE for bringing more people back to the game. I loved HOMM3 when I was a kid and played all the time with family and friends, but as a parent of two kids, I haven't played in years because I no longer have hours of time to sit down away from family to play a synchronous match. However, if I could load a game on my phone, take my turn, go about my day, and get a notification ping when it's next my turn, that would be a huge feature enabling me to play a beloved game that I otherwise wouldn't be able to play. I personally think the way to handle PvP battles would be to temporarily return to synchronous play to finish the battle, then save and return to asynchronous play. Anyway, this is just my two cents. Thank you for all that has been done to keep this game active.

**Comment by trsrm** [comment]
I 100% support the commentator's words above. I recently returned to this game out of nostalgia. I really like the implementation, but the problem is that I simply can't afford to set aside a few hours to play with a friend. In the past month I managed to organize with my friend once and set aside three hours for the game. However, if it were an asynchronous game with notifications, I would have to play constantly and daily. Also attract more players. I would even be willing to pay money for such a feature.

**Comment by IvanSavenko** [comment]
@trsrm 
Not much work has been done here. Any sort of "play by mail" is out of plan. Async MP may be considered. It still has open questions, like how to handle pvp combats, but async MP we might try to just say that players need to decide on time where they are both online

I do have some long-term plans to allow "async MP" or some "resumable game servers"  via our MP lobby, but at the moment I don't have any estimate on when this would happen - still a lot of job to do before it is even possible
